app/Libs/ProductApiDriver.php

```diff
@@ -0,0 +1,150 @@
+        };
+        $response = $this->httpClient->get('/api/product/list?page='.$page.$orderQs);
+
+        return $this->getResponse($response);
+    }
+
+    public function create($data) {
+        $response = $this->httpClient->post('/api/product/create', [
+            'form_params' => [
+                'data' => $data
+            ]
+        ]);
+
+        return $this->getResponse($response);
+    }
+
+    public function update($productId, $data) {
+        $response = $this->httpClient->post('/api/product/update/'.$productId, [
+            'form_params' => [
+                'data' => $data
+            ]
+        ]);
+
+        return $this->getResponse($response);
+    }
+
+    public function search($data, $page = 1) {
+        extract($data);
+        $keyword = isset($keyword) ? $keyword : '';
+	    $categoryId = isset($categoryId) ? $categoryId : 0;
+	    $tag = isset($tag) ? $tag : null;
+	    $priority = isset($priority) ? $priority : null;
+        $orderQs = '';
+        if(isset($orderBy)) {
+            $orderQs = '&orderby='.$orderBy;
+            $orderQs .= '&order='.(isset($order) ? $order : 'desc');
+        }
+        $response = $this->httpClient->get("/api/product/search?keyword={$keyword}&category_id={$categoryId}&tag=".json_encode($tag)."&priority=".json_encode($priority)."&page={$page}{$orderQs}");
+
+        return $this->getResponse($response);
+    }
+
+    public function detail($productId = 0) {
+        $response = $this->httpClient->get('/api/product/detail/'.$productId);
+
+        return $this->getResponse($response);
+    }
+
+    public function url($path) {
+        return $this->apiUrl.$path;
+    }
+
+    public function categories() {
+        $response = $this->httpClient->get("/api/category/list");
+
+        return $this->getResponse($response);
+    }
+
+    public function createCategory($data) {
+        $response = $this->httpClient->post('/api/category/create', [
+            'form_params' => [
+                'data' => $data
+            ]
+        ]);
+
+        return $this->getResponse($response);
+    }
+
+    public function detailCategory($categoryId = 0) {
+        $response = $this->httpClient->get('/api/category/detail/'.$categoryId);
+
+        return $this->getResponse($response);
+    }
+
+    public function updateCategory($categoryId, $data) {
+        $response = $this->httpClient->post('/api/category/update/'.$categoryId, [
+            'form_params' => [
+                'data' => $data
+            ]
+        ]);
+
+        return $this->getResponse($response);
+    }
+
+    public function deleteCategory($categoryId = 0) {
+        $response = $this->httpClient->get('/api/category/delete/'.$categoryId);
+
+        return $this->getResponse($response);
+    }
+
+    public function getAjaxCategories($search = null, $page) {
+        $response = $this->httpClient->get('/api/category/ajaxlist/?search='.$search.'&page='.$page);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "&page=" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1192)